src/ros/mb_publisher.cpp

```diff
@@ -18,6 +18,36 @@ void MBPublisher::publish(ur_msgs::IOStates& io_msg, SharedMasterBoardData& data
   io_pub_.publish(io_msg);
 }
 
+void MBPublisher::publishRobotStatus(SharedRobotModeData& data)
+{
+  industrial_msgs::RobotStatus msg;
+  msg.drives_powered.val = data.robot_power_on;
+  msg.e_stopped.val = data.emergency_stopped;
+  msg.in_motion.val = data.program_running;
+  msg.error_code = 0;
+  msg.in_error.val = data.protective_stopped;
+  msg.mode.val = industrial_msgs::TriState::UNKNOWN;
+
+  status_pub_.publish(msg);
+}
+
+void MBPublisher::publishRobotStatus(RobotModeData_V3_0__1& data)
+{
+  industrial_msgs::RobotStatus msg;
+
+  msg.drives_powered.val = data.robot_power_on;
+  msg.e_stopped.val = data.emergency_stopped;
+  msg.in_motion.val = data.program_running;
+  msg.in_error.val = data.protective_stopped;
```

Comment:
Is `protective_stopped` high for all errors that can occur?

Reply:
there does not seem to be another way to be informed about errors looking at the format in robot_mode.h

Reply:
What about `emergency_stopped`? That would definitely be an 'in_error' situation, but that's not captured here at the moment. ThomasTimm/ur_modern_driver#143 makes `in_error` the union of `protective_stopped` and `emergency_stopped`.

Reply:
ah okay, I thought of e-stopped more like a state and the error would appear separately once you try moving despite it being in e-stopped mode. But I can change it to said union. This would again ask for a place to define the "vendor specific" error codes and then setting `error_code` to differentiate `protective_stopped`

Reply:
"Vendor specific" means that the error codes are defined by the vendor. Not by us, for every vendor.

This field is typically populated by copying the error code that the robot controller gives you for different errors. On Fanucs fi you would be copying the code returned by the `ERR_DATA(..)` Karel function (which is typically a 5 or 6 digit number).

Those codes are defined by Fanuc. We don't make them up ourselves.

UR has similar codes I believe (just take a look at the log on the TP), but I'm not sure we can retrieve those from somewhere.

Reply:
> I thought of e-stopped more like a state and the error would appear separately once you try moving despite it being in e-stopped mode.

That could definitely be a way to implement it, but we would need a way to reliably determine which of the two it is. I don't have a scripting manual with me right now, but if there is a `get_last_error_code()` function that we can use, then we can definitely do what you propose.

Reply:
You're correct. `e_stopped` is the state, and `in_error` should follow from whether the robot controller defines this as an error.

Let's keep what you have for now.

Sorry for the noise.

Reply:
re: vendor specific codes: looking at a UR we have here I keep seeing `C100A5` in the log whenever the robot is `Robot changed mode: EMERGENCY STOPPED`.

Those are the codes that would go into `error_code`.